File: BitManipulation/SingleNum2.java
public class SingleNum2 {
    public static int singleNumber(int[] nums) {
        int ans = 0;
        for(int i = 0; i < 32; i++) {
            int sum = 0;
            for (int num : nums) {
                if (((num >> i) & 1) == 1) {
                    sum++;
                    sum %= 3;
                }

            }
            if(sum != 0) {
                ans |= sum << i;
            }
            System.out.println(ans);
        }
        return ans;
    }

    public static void main(String[] args) {
        System.out.println(singleNumber(new int[]{1, 2, 2, 2, 1, 1, 3}));
    }
}

File: BitManipulation/SingleNum2.py
def singleNumber(nums):
    seenOnce, seenTwice = 0, 0

    for num in nums:
        # first appearence: 
        # add num to seen_once 
        # don't add to seen_twice because of presence in seen_once

        # second appearance: 
        # remove num from seen_once 
        # add num to seen_twice

        # third appearance: 
        # don't add to seen_once because of presence in seen_twice
        # remove num from seen_twice
        seenOnce = ~seenTwice & (seenOnce ^ num)
        seenTwice = ~seenOnce & (seenTwice ^ num)
    

    return seenOnce

print(singleNumber([1, 2, 3, 1, 2, 1, 2]))